File: apps/go-api/domains/applications/applications.go
package applications

import "time"

type Application struct {
	id            int64
	name          string
	env           string
	bundleJS      string
	styleURL      string
	remoteOrigin  string
	proxyBasePath string
	basePath      string
	createdAt     time.Time
	updatedAt     time.Time
}

func (a *Application) ID() int64             { return a.id }
func (a *Application) Name() string          { return a.name }
func (a *Application) Env() string           { return a.env }
func (a *Application) BundleJS() string      { return a.bundleJS }
func (a *Application) StyleURL() string      { return a.styleURL }
func (a *Application) RemoteOrigin() string  { return a.remoteOrigin }
func (a *Application) ProxyBasePath() string { return a.proxyBasePath }
func (a *Application) BasePath() string      { return a.basePath }
func (a *Application) CreatedAt() time.Time  { return a.createdAt }
func (a *Application) UpdatedAt() time.Time  { return a.updatedAt }

func NewApplication(
	id int64,
	name, env, bundleJS, styleURL, remoteOrigin, proxyBasePath, basePath string,
	createdAt, updatedAt time.Time,
) *Application {
	return &Application{
		id:            id,
		name:          name,
		env:           env,
		bundleJS:      bundleJS,
		styleURL:      styleURL,
		remoteOrigin:  remoteOrigin,
		proxyBasePath: proxyBasePath,
		basePath:      basePath,
		createdAt:     createdAt,
		updatedAt:     updatedAt,
	}
}

File: packages/mfe-host/src/queries/applications.ts
import { useEffect, useState } from "react";

export interface ApplicationConfig {
  name: string;
  bundleJs?: string;
  styleUrl?: string;
  remoteOrigin?: string;
  proxyBasePath?: string;
  basePath?: string;
}

interface ApiApplicationRow {
  id: number;
  name: string;
  env: string;
  bundle_js: string;
  style_url?: string;
  remote_origin: string;
  proxy_base_path: string;
  base_path: string;
}

interface ApiApplicationsResponse {
  data: ApiApplicationRow[];
}

export interface UseApplicationsResult {
  data: ApplicationConfig[] | null;
  loading: boolean;
  error: Error | null;
}

function toApplicationConfig(row: ApiApplicationRow): ApplicationConfig {
  return {
    name: row.name,
    bundleJs: row.bundle_js,
    styleUrl: row.style_url,
    remoteOrigin: row.remote_origin,
    proxyBasePath: row.proxy_base_path,
    basePath: row.base_path,
  };
}

/**
 * Dependency-free registry hook: plain fetch + useState/useEffect.
 * Reads the public go-api registry (GET /api/applications, proxied same-origin).
 */
export function useApplications(): UseApplicationsResult {
  const [data, setData] = useState<ApplicationConfig[] | null>(null);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState<Error | null>(null);

  useEffect(() => {
    const controller = new AbortController();
    let cancelled = false;

    fetch("/api/applications", { signal: controller.signal })
      .then((res) => {
        if (!res.ok) {
          throw new Error(`Failed to fetch applications: HTTP ${res.status}`);
        }
        return res.json() as Promise<ApiApplicationsResponse>;
      })
      .then((body) => {
        if (cancelled) {
          return;
        }
        setData((body.data ?? []).map(toApplicationConfig));
        setLoading(false);
      })
      .catch((err: unknown) => {
        if (cancelled) {
          return;
        }
        setError(err instanceof Error ? err : new Error(String(err)));
        setLoading(false);
      });

    return () => {
      cancelled = true;
      controller.abort();
    };
  }, []);

  return { data, loading, error };
}
